src/parcel_snoopi.jl

```diff
@@ -380,3 +380,150 @@ end
 const default_exclusions = Set([
     r"\bMain\b",
 ])
+
+# === @snoopi_deep helper functions ========================================================
+
+"""
+    flatten_times(timing::Core.Compiler.Timings.Timing; tmin_secs = 0.0)
+
+Flatten the execution graph of Timings returned from `@snoopi_deep` into a Vector of pairs,
+with the exclusive time for each invcation of type inference within the compiler, sorted by
+the exclusive time, skipping any frames that took less than `tmin_secs` seconds.
+"""
+function flatten_times(timing::Core.Compiler.Timings.Timing; tmin_secs = 0.0)
+    out = Any[]
+    frontier = [timing]
+    while !isempty(frontier)
+        t = popfirst!(frontier)
+        exclusive_time = (t.time / 1e9)
+        if exclusive_time >= tmin_secs
+            push!(out, exclusive_time => t.mi_info)
+        end
+        for c in t.children
+            push!(frontier, c)
+        end
+    end
+    return sort(out; by=tl->tl[1])
+end
+
+
+import FlameGraphs
+
+using Base.StackTraces: StackFrame
+using LeftChildRightSiblingTrees: Node, addchild
+using Core.Compiler.Timings: Timing
+
+struct InclusiveTiming
+    mi_info::Core.Compiler.Timings.InferenceFrameInfo
+    inclusive_time::UInt64
+    start_time::UInt64
```

Comment:
What is `start_time` used for? (More about this below.)

Reply:
This is needed because the nodes in `FlameGraph` aren't just a duration: they are a _range_, from a start sample to an end sample:
```julia
julia> fg = SnoopCompile.to_flamegraph(timing)
Node(FlameGraphs.NodeData(ROOT() at typeinfer.jl:70, 0x00, 0:14424075))

julia> fg.child.child.child
Node(FlameGraphs.NodeData(i(x::Integer) at REPL[8]:2, 0x00, 168721:646740))
```
This (`0:14424075` vs `168721:646740`) is what allows the profile to have elements show up at the actual time they occurred, relative to eachother.

For example, in this image, `i(x::Int64)` only occurs at _runtime_, via dynamic dispatch, so you see a big gap in-between the first inference stack and it:

<img width="400" alt="94317053-7fd11a00-ff53-11ea-9f7a-7e5ba7de0eef" src="https://user-images.githubusercontent.com/1582097/97182670-34d53d00-1773-11eb-8fa2-3cb0b6406eb0.png">

Which comes from this:
```julia
julia> fg.child.sibling
Node(FlameGraphs.NodeData(i(x::Int64) at REPL[8]:2, 0x00, 11762162:11931975))
```

Unless I'm missing something, this `start_time` logic is required to get that to work, because in each node in julia's `Core.Compiler.Timings.Timing`, we're only keeping the total exclusive time and the start time:
https://github.com/NHDaly/julia/blob/f6f4705f943abebcb4eaa50bb78e9cb9f46d0c43/base/compiler/typeinfer.jl#L48-L56

So to recompute the spans needed for the `range` in the FlameGraph, we need to keep the start time, and then sum the inclusive time with the inclusive times of all children, which gives the range, as `t.start_time : (t.start_time + t.inclusive_time)`.

Does that answer your question? :)

Reply:
Very nicely, thanks!